test/e2e/storage/testsuites/snapshottable.go

```diff
@@ -174,29 +173,30 @@ func (s *snapshottableTestSuite) DefineTests(driver storageframework.TestDriver,
 			//     - Windows: data is flushed to disk and the disk is detached
 			//
 			// Pod deletion might not guarantee a data flush to disk, however NodeUnstageVolume adds the logic
-			// to flush the data to disk (see #81690 for details). We need to wait for NodeUnstageVolume, as
+			// to flush the data to disk (see #81690 for details).
 			// NodeUnpublishVolume only removes the bind mount, which doesn't force the caches to flush.
 			// It's possible to create empty snapshots if we don't wait (see #101279 for details).
 			//
-			// In the following code by checking if the PV is not in the node.Status.VolumesInUse field we
-			// ensure that the volume is not used by the node anymore (an indicator that NodeUnstageVolume has
-			// already finished)
+			// Because of this possibility, we need to wait for the VolumeAttachment to be deleted, as
+			// this ensures that NodeUnstageVolume has completed, and the cache has been flushed.
+			//
+			// In the following code we check to ensure that the VolumeAttachent no longer exists
+			// before proceeding.
 			nodeName := pod.Spec.NodeName
 			gomega.Expect(nodeName).NotTo(gomega.BeEmpty(), "pod.Spec.NodeName must not be empty")
 
-			ginkgo.By(fmt.Sprintf("[init] waiting until the node=%s is not using the volume=%s", nodeName, pv.Name))
+			ginkgo.By(fmt.Sprintf("[init] waiting until the VolumeAttachment is removed"))
 			success := storageutils.WaitUntil(framework.Poll, f.Timeouts.PVDelete, func() bool {
-				node, err := cs.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
-				framework.ExpectNoError(err)
-				volumesInUse := node.Status.VolumesInUse
-				framework.Logf("current volumes in use: %+v", volumesInUse)
-				for i := 0; i < len(volumesInUse); i++ {
-					if strings.HasSuffix(string(volumesInUse[i]), pv.Name) {
-						return false
-					}
+				volumeAttachment, err := storageutils.GetVolumeAttachment(cs, pv, sc.Provisioner, nodeName)
```

Comment:
I think you need to wait for both - volume not in `volumesInUse` (for non-attachable volumes) and for missing VolumeAttachment (for attachable ones). A comment about AWS not flushing its caches on NodeStage would be great :-).

Reply:
This has been updated to include the original wait as well as the new one.